k8s-test-runner/k8s_client/client.go

```diff
@@ -0,0 +1,262 @@
+			TimeoutSeconds: &timeout,
+		})
+		return err
+	}
+
+	err := retryK8sCall(call, m.callRetryPolicy)
+	if err != nil {
+		// Wrap and return any error encountered during the retry operation
+		return nil, fmt.Errorf(`failed to call CoreV1().Pods().List(), namespace: %s, labelSelector: %s, timeout: %d: %w`, namespace, labelSelector, timeout, err)
+	}
+
+	// At this point, `pods` should be populated successfully
+	return pods, nil
+}
+
+func (m *Client) ListJobs(ctx context.Context, namespace, syncLabel string) (*batchV1.JobList, error) {
+	var jobs *batchV1.JobList
+
+	timeout := int64(30)
+	labelSelector := syncSelector(syncLabel)
+	call := func() error {
+		var err error
+		jobs, err = m.ClientSet.BatchV1().Jobs(namespace).List(ctx, metaV1.ListOptions{
+			LabelSelector:  labelSelector,
+			TimeoutSeconds: &timeout,
+		})
+		return err
+	}
+
+	err := retryK8sCall(call, m.callRetryPolicy)
+	if err != nil {
+		// Wrap and return any error encountered during the retry operation
+		return nil, fmt.Errorf(`failed to call BatchV1().Jobs().List(), namespace: %s, labelSelector: %s, timeout: %d: %w`, namespace, labelSelector, timeout, err)
+	}
+
+	// At this point, `jobs` should be populated successfully
+	return jobs, nil
+}
+
+func (m *Client) GetPodLogs(ctx context.Context, namespace string, pods []v1.Pod) (map[string]string, error) {
+	podLogs := make(map[string]string)
+
+	call := func() error {
+		for _, pod := range pods {
+			req := m.ClientSet.CoreV1().Pods(namespace).GetLogs(pod.Name, &v1.PodLogOptions{})
+			podLog, err := req.Stream(ctx)
+			if err != nil {
+				return fmt.Errorf("failed to open log stream for pod %s: %w", pod.Name, err)
+			}
+			defer podLog.Close()
+
+			logs, err := io.ReadAll(podLog)
+			if err != nil {
+				return fmt.Errorf("failed to read log for pod %s: %w", pod.Name, err)
+			}
+
+			podLogs[pod.Name] = string(logs)
+		}
+
+		return nil // Success
+	}
+
+	err := retryK8sCall(call, m.callRetryPolicy)
+	if err != nil {
+		// Wrap and return any error encountered during the retry operation
+		return nil, fmt.Errorf("failed to retrieve pod logs: %w", err)
+	}
+
+	return podLogs, nil
+}
+
+func syncSelector(s string) string {
+	return fmt.Sprintf("sync=%s", s)
+}
+
+func (m *Client) RemoveJobs(ctx context.Context, nsName string, syncLabelValue string) error {
+	jobs, err := m.ListJobs(ctx, nsName, syncLabelValue)
+	if err != nil {
+		return err
+	}
+
+	log.Info().Interface("jobs", jobs).Msg("Removing jobs")
+
+	for _, j := range jobs.Items {
+		dp := metaV1.DeletePropagationForeground
+		if err := m.ClientSet.BatchV1().Jobs(nsName).Delete(ctx, j.Name, metaV1.DeleteOptions{
+			PropagationPolicy: &dp,
+		}); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Use polling to wait for jobs to complete.
+// `watch` does not work in CI (GAP), so we have to poll.
+func (m *Client) WaitUntilJobsComplete(ctx context.Context, namespace, syncLabelValue string, expectedJobsCount int) error {
```

Comment:
That's because we need to support every k8s call in GAP explicitly and there is no implementation, right?